src/audio/chain_dma.c

```diff
@@ -37,56 +48,478 @@ static struct comp_dev *chain_dma_create(const struct comp_driver *drv,
+	dma_release_channel(cd->chan_link->dma->z_dev, cd->chan_link->index);
+
+	if (cd->dma_buffer) {
+		buffer_free(cd->dma_buffer);
+		cd->dma_buffer = NULL;
+	}
+}
+
+int chain_task_init(struct comp_dev *dev, uint8_t host_dma_id, uint8_t link_dma_id,
+		    uint32_t fifo_size)
+{
+	struct comp_buffer __sparse_cache *buffer_c;
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	uint32_t dir;
+	int ret;
+
+	schedule_task_init_ll(&cd->chain_task,
+			      SOF_UUID(chain_dma_uuid),
+			      SOF_SCHEDULE_LL_TIMER, SOF_TASK_PRI_LOW,
+			      chain_task_run, cd, 0, 0);
+
+	ret = get_connector_node_id(host_dma_id, true, &cd->host_connector_node_id);
+	if (ret)
+		return ret;
+
+	ret = get_connector_node_id(link_dma_id, false, &cd->link_connector_node_id);
+	if (ret)
+		return ret;
+
+	/* Verify whether HDA gateways can be chained */
+	if (cd->host_connector_node_id.f.dma_type == ipc4_hda_host_output_class) {
+		if (!(cd->link_connector_node_id.f.dma_type == ipc4_hda_link_output_class))
+			return IPC4_ERROR_INVALID_PARAM;
+		cd->stream_direction = SOF_IPC_STREAM_PLAYBACK;
+	}
+	if (cd->host_connector_node_id.f.dma_type == ipc4_hda_host_input_class) {
+		if (!(cd->link_connector_node_id.f.dma_type == ipc4_hda_link_input_class))
+			return IPC4_ERROR_INVALID_PARAM;
+		cd->stream_direction = SOF_IPC_STREAM_CAPTURE;
+	}
+
+	/* request HDA DMA with shared access privilege */
+	dir = (cd->stream_direction == SOF_IPC_STREAM_PLAYBACK) ?
+		DMA_DIR_HMEM_TO_LMEM : DMA_DIR_LMEM_TO_HMEM;
+
+	cd->dma_host = dma_get(dir, 0, DMA_DEV_HOST, DMA_ACCESS_SHARED);
+
+	if (!cd->dma_host) {
+		comp_err(dev, "chain_task_init(): dma_get() returned NULL");
+		return IPC4_FAILURE;
+	}
+
+	dir = (cd->stream_direction == SOF_IPC_STREAM_PLAYBACK) ?
+	DMA_DIR_MEM_TO_DEV : DMA_DIR_DEV_TO_MEM;
+
+	cd->dma_link = dma_get(dir, DMA_CAP_HDA, DMA_DEV_HDA, DMA_ACCESS_SHARED);
+
+	if (!cd->dma_link) {
+		comp_err(dev, "chain_task_init(): dma_get() returned NULL");
+		return IPC4_FAILURE;
+	}
+
+	/* retrieve DMA buffer address alignment */
+	uint32_t addr_align;
+	int err = dma_get_attribute(cd->dma_host, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT, &addr_align);
+
+	if (err < 0) {
+		comp_err(dev, "chain_task_init(): could not get dma buffer address alignment, err = %d",
+			 err);
+		return IPC4_FAILURE;
+	}
+
+	fifo_size = ALIGN_UP_INTERNAL(fifo_size, addr_align);
+
+	if (cd->link_connector_node_id.f.dma_type == ipc4_hda_link_input_class) {
+		/* Increasing buffer size for capture path as L1SEN exit takes sometimes
+		 * more than expected.
+		 */
+		fifo_size *= 5;
+	}
+
+	if (cd->link_connector_node_id.f.dma_type == ipc4_hda_link_output_class) {
+		/* Increasing buffer size for playback path as L1SEN exit takes sometimes
+		 * more that expected
+		 * Note, FIFO size must be smaller than half of host buffer size
+		 * (20ms ping pong) to avoid problems with position reporting
+		 * Size increase from default 2ms to 5ms is enough.
+		 */
+		fifo_size /= 2;
+		fifo_size *= 5;
+	}
```

Comment:
if it is
`cd->link_connector_node_id.f.dma_type != ipc4_hda_link_input_class && cd->link_connector_node_id.f.dma_type != ipc4_hda_link_output_class` then the fifo_size does not need to be modified and it is valid?

Reply:
Only 2 classes of dma link type are possible:  ipc4_hda_link_input_class or ipc4_hda_link_output_class. It is always being modified. Validation node id and type assignment is realized in get_connector_node_id(). I replaced it with switch. Is it more clear ?

Reply:
@ujfalusi resolved ?